File: components/webfinger.go
// Webfinger component to demo static JSON
package main

import (
	_ "embed"
	"net/http"
	"os"

	spinhttp "github.com/fermyon/spin/sdk/go/http"
)

//go:embed webfinger.json
var webfingerJSON []byte

func init() {
	spinhttp.Handle(webfingerHand)
}

func main() {}

func webfingerHand(w http.ResponseWriter, r *http.Request) {
	if r.URL.Path != "/well-known/webfinger" &&
		r.URL.Path != "/.well-known/webfinger" {
		http.NotFound(w, r)
		return
	}
	if r.Method != http.MethodGet {
		http.NotFound(w, r)
		return
	}
	var resource = r.URL.Query().Get("resource")
	if resource == "" {
		http.Error(w, "The resource query parameter is missing", http.StatusBadRequest)
		return
	}

	if unknownResource(resource) {
		http.Error(w, "Unknown resource", http.StatusBadRequest)
		return
	}

	w.Header().Set("Content-Type", "application/jrd+json")
	w.Header().Set("Access-Control-Allow-Origin", "*")

	w.Write(webfingerJSON)
}

func unknownResource(resource string) bool {
	var match = false
	var re = formatResource()
	for _, known := range re {
		if resource == known {
			match = true
			break
		}
	}
	return !match
}
func formatResource() []string {
	var self = os.Getenv("SELF_ACTOR")
	var subd = os.Getenv("SITE_NAME")
	var ar []string

	//case "acct:self@subd":
	ar = append(ar, "acct:"+self+"@"+subd+".fermyon.app")

	//case "mailto:self@subd"
	ar = append(ar, "mailto:"+self+"@"+subd+".fermyon.app")

	//case "https://subd"
	ar = append(ar, "https:"+subd+".fermyon.app")

	//case "https://subd/"
	ar = append(ar, "https:"+subd+".fermyon.app/")

	return ar
}

File: components/skeleton-zig/src/webfinger.zig
const std = @import("std");
const lib = @import("lib.zig");
const str = @import("strings.zig");
const status = @import("status.zig");
const config = @import("config.zig");
const Allocator = std.mem.Allocator;
const log = std.log;

const Webfinger = @This();

const Impl = WebfingerImpl;

pub fn eval(allocator: Allocator, w: *lib.HttpResponse, r: *lib.HttpRequest) void {
    Impl.eval(allocator, w, r);
}

const webfinger_json = @embedFile("webfinger.json");

const WebfingerImpl = struct {
    fn eval(allocator: Allocator, w: *lib.HttpResponse, req: *lib.HttpRequest) void {
        if (req.method != 0) return status.nomethod(w);

        const unknown = unknownResource(allocator, req.uri);
        if (unknown) return status.bad(w);

        w.headers.put("Content-Type", "application/jrd+json") catch {
            log.err("ERROR response header", .{});
        };
        w.headers.put("Access-Control-Allow-Origin", "*") catch {
            log.err("ERROR response header", .{});
        };
        w.body.appendSlice(webfinger_json) catch {
            log.err("ERROR webfinger body", .{});
            return status.internal(w);
        };

        status.ok(w);
    }
};

// check query param 'resource'
fn unknownResource(allocator: Allocator, ur: []const u8) bool {
    const bad = true;
    var map = str.qryParams(allocator, ur);
    defer map.deinit();

    var resource: []const u8 = undefined;
    if (map.get("resource")) |val| {
        resource = val;
    } else {
        log.err("param resource is required", .{});
        return bad;
    }
    var grp = formatResource(allocator) catch {
        log.err("resource list OutOfMem", .{});
        return bad;
    };
    defer grp.deinit();
    const allowed = grp.items;
    const decoded = str.percentDecode(allocator, resource) catch {
        log.err("decode OutOfMem", .{});
        return bad;
    };
    defer allocator.free(decoded);
    for (allowed) |known| {
        log.debug("cmp {s} ({d}) to {s} ({d})", .{ known, known.len, decoded, decoded.len });

        if (std.mem.eql(u8, known, decoded)) {
            return !bad;
        }
    }

    return bad;
}

// list allowed resource values
fn formatResource(allocator: Allocator) !std.ArrayList([]const u8) {
    var all = std.ArrayList([]const u8).init(allocator);
    errdefer all.deinit();

    const who = config.selfActor() orelse "00000";
    const subd = config.siteSubdomain() orelse "00000";

    //case "acct:self@subd":
    const c1 = try std.fmt.allocPrint(allocator, "acct:{s}@{s}", .{ who, subd });
    try all.append(c1);

    //case "mailto:self@subd"
    const c2 = try std.fmt.allocPrint(allocator, "mailto:{s}@{s}", .{ who, subd });
    try all.append(c2);

    //case "https://subd"
    const c3 = try std.fmt.allocPrint(allocator, "https://{s}", .{subd});
    try all.append(c3);

    //case "https://subd/"
    const c4 = try std.fmt.allocPrint(allocator, "https://{s}/", .{subd});
    try all.append(c4);

    return all;
}
